presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -223,11 +224,22 @@ public PlanWithProperties visitOutput(OutputNode node, Context context)
         {
             PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.undistributed()));
 
+            // if the child is already a gathering exchange, don't add another exchange
+            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType() == ExchangeNode.Type.GATHER) {
+                return rebaseAndDeriveProperties(node, child);
+            }
+
             if (!child.getProperties().isSingleNode()) {
                 child = withDerivedProperties(
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else if (!child.getProperties().isCoordinatorOnly()) {
```

Comment:
Here, the check `isCoordinatorOnly()` is done on `child` node and *not* the output node. So we do need this check.